src/Xamarin.Android.Build.Tasks/Utilities/XAAssemblyResolver.cs

```diff
@@ -0,0 +1,333 @@
+			}
+		}
+
+		void Dispose (bool disposing)
+		{
+			if (disposed || !disposing) {
+				return;
+			}
+
+			Default = null;
+			foreach (var kvp in assemblies) {
+				kvp.Value?.Dispose ();
+			}
+			assemblies.Clear ();
+			disposed = true;
+		}
+
+		public void Dispose ()
+		{
+			Dispose (disposing: true);
+			GC.SuppressFinalize (this);
+		}
+	}
+
+	/// <summary>
+	/// Contains a collection of directories where framework assemblies can be found.  This collection **must not**
+	/// contain any directories which contain ABI-specific assemblies.  For those, use <see cref="AbiSearchDirectories"/>
+	/// </summary>
+	public ICollection<string> FrameworkSearchDirectories { get; } = new List<string> ();
+
+	/// <summary>
+	/// Contains a collection of directories where Xamarin.Android (via linker, for instance) has placed the ABI
+	/// specific assemblies.  Each ABI has its own set of directories to search.
+	/// <summary>
+	public IDictionary<AndroidTargetArch, ICollection<string>> AbiSearchDirectories { get; } = new Dictionary<AndroidTargetArch, ICollection<string>> ();
+
+	readonly List<MemoryMappedViewStream> viewStreams = new List<MemoryMappedViewStream> ();
+	bool disposed;
+	TaskLoggingHelper log;
+	bool loadDebugSymbols;
+	ReaderParameters readerParameters;
+	readonly Dictionary<string, CacheEntry> cache;
+
+	public XAAssemblyResolver (TaskLoggingHelper log, bool loadDebugSymbols, ReaderParameters? loadReaderParameters = null)
+	{
+		this.log = log;
+		this.loadDebugSymbols = loadDebugSymbols;
+		this.readerParameters = loadReaderParameters ?? new ReaderParameters();
+
+		cache = new Dictionary<string, CacheEntry> (StringComparer.OrdinalIgnoreCase);
+	}
+
+	public AssemblyDefinition Resolve (string fullName, ReaderParameters? parameters = null)
+	{
+		return Resolve (AssemblyNameReference.Parse (fullName), parameters);
+	}
+
+	public AssemblyDefinition Resolve (AssemblyNameReference name)
+	{
+		return Resolve (name, null);
+	}
+
+	public AssemblyDefinition Resolve (AssemblyNameReference name, ReaderParameters? parameters)
+	{
+		return Resolve (AndroidTargetArch.None, name, parameters);
+	}
+
+	public AssemblyDefinition Resolve (AndroidTargetArch arch, AssemblyNameReference name, ReaderParameters? parameters = null)
+	{
+		string shortName = name.Name;
+		if (cache.TryGetValue (shortName, out CacheEntry? entry)) {
+			return SelectAssembly (arch, name.FullName, entry, loading: false);
+		}
+
+		if (arch == AndroidTargetArch.None) {
+			return FindAndLoadFromDirectories (arch, FrameworkSearchDirectories, name, parameters);
+		}
+
+		if (!AbiSearchDirectories.TryGetValue (arch, out ICollection<string>? directories) || directories == null) {
+			throw CreateLoadException (name);
+		}
+
+		return FindAndLoadFromDirectories (arch, directories, name, parameters);
+	}
+
+	AssemblyDefinition FindAndLoadFromDirectories (AndroidTargetArch arch, ICollection<string> directories, AssemblyNameReference name, ReaderParameters? parameters)
```

Comment:
Why have `ReaderParameters? parameters` if you're not going to use it?  Shouldn't the call to `Load(arch, assemblyFile)` provide `parameters`?